### PR TITLE
adds support for editor parameter in tokens

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- adds support for editor parameter in tokens

## Why are we making these changes?

Allows giving editing privileges to another account other than the owner.

### CODE REVIEW

File: waiter/src/waiter/token.clj [hunk 1]
```diff
@@ -516,48 +516,61 @@
...
 
       nil
-      (do
+      (let [existing-editor (get existing-token-metadata "editor")
+            existing-owner (get existing-token-metadata "owner")
+            owner? (and existing-owner (authz/manage-token? entitlement-manager authenticated-user token existing-token-metadata))
+            editor? (and existing-editor (not owner?)  (authz/run-as? entitlement-manager authenticated-user existing-editor))
+            delegated-user (if editor?
+                             (do ;; allow editor to run with owner privileges
+                               (log/info "promoting editor to owner privileges while editing token"
```
Comment (sradack) [inline waiter/src/waiter/token.clj:525]: Can we update this log message?  Editors don't have the same permissions as owners, so the editor doesn't actually have owner privileges.

File: waiter/src/waiter/token.clj [hunk 2]
```diff
@@ -516,48 +516,61 @@
                            :log-level :warn}))))
 
       nil
-      (do
+      (let [existing-editor (get existing-token-metadata "editor")
+            existing-owner (get existing-token-metadata "owner")
```
Comment (sradack) [inline waiter/src/waiter/token.clj:520]: We used to have code that allowed for tokens to not have an existing owner.  Does this break that backward compatibility?

File: waiter/src/waiter/token.clj [hunk 3]
```diff
@@ -516,48 +516,61 @@
...
+            owner? (and existing-owner (authz/manage-token? entitlement-manager authenticated-user token existing-token-metadata))
+            editor? (and existing-editor (not owner?)  (authz/run-as? entitlement-manager authenticated-user existing-editor))
+            delegated-user (if editor?
```
Comment (sradack) [inline waiter/src/waiter/token.clj:523]: A weird situation can happen when setting an editor on a token that doesn't have an owner.  e.g. If run-as-user is set to me, and owner is null, and I set editor to me as well, I can no longer change the owner, editor, or run-as-user.

File: waiter/src/waiter/token.clj [hunk 4]
```diff
@@ -516,48 +516,61 @@
...
+                                  {:authenticated-user authenticated-user
+                                   :existing-token-description existing-token-parameters
+                                   :parameter parameter-name
+                                   :parameter-exiting-value existing-value
+                                   :parameter-new-value new-value
+                                   :status http-403-forbidden
+                                   :log-level :warn})))))))
         (when (and run-as-user (not= "*" run-as-user))
-          (when-not (authz/run-as? entitlement-manager authenticated-user run-as-user)
+          (when-not (authz/run-as? entitlement-manager delegated-user run-as-user)
```
Comment (sradack) [inline waiter/src/waiter/token.clj:544]: Previously we were checking if the authenticated user could `run-as?`. Now we're checking if the owner can `run-as?` It's not clear why we are changing the original behavior.

File: waiter/src/waiter/token.clj [hunk 5]
```diff
@@ -516,48 +516,61 @@
...
             (throw (ex-info "Cannot run as user"
                             {:authenticated-user authenticated-user
+                             :delegated-user delegated-user
```
Comment (sradack) [inline waiter/src/waiter/token.clj:547]: The `:delegated-user` key and `delegated-user` symbol are named curiously in the case of editing as an owner.  If I'm editing the token as the owner, I'm not delegating.

**Review by sradack — CHANGES_REQUESTED** [review]

File: waiter/src/waiter/token.clj (hunk 2 above)
Comment (sradack) [reply waiter/src/waiter/token.clj:520]: What if a token doesn't have a `run-as-user` or `run-as-user` is `*` or ...

File: waiter/src/waiter/token.clj (hunk 3 above)
Comment (sradack) [reply waiter/src/waiter/token.clj]: Do we always have a `run-as-user`?

File: waiter/src/waiter/token.clj [hunk 6]
```diff
@@ -516,48 +516,62 @@
...
-          (throw (ex-info "Cannot modify root token metadata"
-                          {:status http-400-bad-request
-                           :token-metadata new-token-metadata
-                           :log-level :warn})))
-        (when (contains? new-token-metadata "previous")
-          (throw (ex-info "Cannot modify previous token metadata"
-                          {:status http-400-bad-request
-                           :token-metadata new-token-metadata
-                           :log-level :warn}))))
+        (if-not (str/blank? existing-owner)
```
Comment (sradack) [inline waiter/src/waiter/token.clj:549]: Isn't your claim that `existing-owner` is always available?

File: waiter/src/waiter/token.clj [hunk 7]
```diff
@@ -516,48 +516,62 @@
...
+                                  {:authenticated-user authenticated-user
+                                   :existing-token-description existing-token-parameters
+                                   :parameter parameter-name
+                                   :parameter-exiting-value existing-value
+                                   :parameter-new-value new-value
+                                   :privileges {:editor? editor? :owner? owner?}
+                                   :status http-403-forbidden
+                                   :log-level :warn}))))))
+          ;; only check run-as-user rules when not running as editor
+          (when (and run-as-user (not= "*" run-as-user))
```
Comment (sradack) [inline waiter/src/waiter/token.clj:541]: Nitpick: I think it reads more cleanly if move this out of the else clause into it's own high level "rule".  It's unrelated to the branch taken in the `true` case.

File: waiter/src/waiter/token.clj [hunk 8]
```diff
@@ -516,48 +516,62 @@
...
+            (when-not (authz/run-as? entitlement-manager authenticated-user run-as-user)
+              (throw (ex-info "Cannot run as user"
                               {:authenticated-user authenticated-user
-                               :owner owner
+                               :privileges {:editor? editor? :owner? owner?}
```
Comment (sradack) [inline waiter/src/waiter/token.clj:545]: Can `editor?` ever be anything other than `false` here?  Should we revert back to the old log message?

File: waiter/src/waiter/token.clj (hunk 6 above)
Comment (sradack) [reply waiter/src/waiter/token.clj:549]: If this check for existing owner is really checking whether the token exists or not, can we make that a bit more explicit?

File: waiter/src/waiter/token.clj (hunk 2 above)
Comment (shamsimam) [reply waiter/src/waiter/token.clj:520]: Oops!

I changed the code to account for a missing owner. This should not break backward compatibility now.

> A weird situation can happen when setting an editor on a token that doesn't have an owner. e.g. If run-as-user is set to me, and owner is null, and I set editor to me as well, I can no longer change the owner, editor, or run-as-user.
> https://github.com/twosigma/waiter/pull/1313#discussion_r602824305

Newly created/modified tokens always have an owner. A token with an editor, will not have a `null` owner.

File: waiter/src/waiter/token.clj (hunk 3 above)
Comment (shamsimam) [reply waiter/src/waiter/token.clj]: Responded here: https://github.com/twosigma/waiter/pull/1313#discussion_r603327556

File: waiter/src/waiter/token.clj [hunk 9]
```diff
@@ -516,48 +516,65 @@
...
-                          {:status http-400-bad-request
-                           :token-metadata new-token-metadata
-                           :log-level :warn}))))
+                               :log-level :warn}))))
+          ;; new token creation
+          (when-not (authz/run-as? entitlement-manager authenticated-user owner)
+            (throw (ex-info "Cannot create token as user"
+                            {:authenticated-user authenticated-user
+                             :owner owner
+                             :privileges {:editor? editor? :owner? owner?}
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/token.clj]: Removed the `privileges` section from the map to return to the state where we were before introducing editor changes.

File: waiter/src/waiter/token.clj [hunk 10]
```diff
@@ -516,48 +516,65 @@
                            :log-level :warn}))))
 
       nil
-      (do
-        (when (and run-as-user (not= "*" run-as-user))
+      (let [existing-editor (get existing-token-metadata "editor")
+            existing-owner (get existing-token-metadata "owner")
+            creating-token? (empty? existing-token-metadata)
+            owner? (and existing-owner
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/token.clj]: Renamed to `existing-owner?` to avoid confusion.

File: waiter/src/waiter/token.clj [hunk 11]
```diff
@@ -516,48 +516,64 @@
                            :log-level :warn}))))
 
       nil
-      (do
-        (when (and run-as-user (not= "*" run-as-user))
+      (let [existing-editor (get existing-token-metadata "editor")
+            existing-owner (get existing-token-metadata "owner")
+            creating-token? (empty? existing-token-metadata)
+            existing-owner? (and existing-owner
```
Comment (sradack) [inline waiter/src/waiter/token.clj:522]: Nit-pick: One might confuse the meaning of `existing-owner?` with the meaning "token has an existing owner".  Here it means "token has an existing owner and the authenticated-user can be that owner"

**Review by sradack — CHANGES_REQUESTED** [review]

File: waiter/src/waiter/token.clj (hunk 11 above)
Comment (shamsimam) [reply waiter/src/waiter/token.clj:522]: Changed to `current-owner?`

File: waiter/src/waiter/token.clj [hunk 12]
```diff
@@ -516,48 +516,64 @@
...
+                                 (authz/manage-token? entitlement-manager authenticated-user token existing-token-metadata))
+            editor? (and (not creating-token?)
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/token.clj:524]: Renamed to `editing?`

File: waiter/src/waiter/token.clj [hunk 13]
```diff
@@ -516,48 +516,64 @@
...
           (when-not (authz/run-as? entitlement-manager authenticated-user run-as-user)
             (throw (ex-info "Cannot run as user"
                             {:authenticated-user authenticated-user
                              :run-as-user run-as-user
                              :status http-403-forbidden
                              :log-level :warn}))))
-        (let [existing-service-description-owner (get existing-token-metadata "owner")]
-          (if-not (str/blank? existing-service-description-owner)
-            (when-not (authz/manage-token? entitlement-manager authenticated-user token existing-token-metadata)
+        (if-not creating-token?
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/token.clj:552]: Swapped the clauses, I hadn't done so earlier to make the diff easier to read.

File: waiter/src/waiter/token.clj [hunk 14]
```diff
@@ -516,48 +516,64 @@
...
+          ;; editing token
+          (let [delegated-user (or (when editor? existing-owner) authenticated-user)]
+            (when-not (authz/manage-token? entitlement-manager delegated-user token existing-token-metadata)
               (throw (ex-info "Cannot change owner of token"
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply waiter/src/waiter/token.clj:556]: Changed to `Cannot update token`

**Review by sradack — APPROVED** [review]

**Pull request merged**